app/models/reports/repository.rb

```diff
@@ -0,0 +1,100 @@
+module Reports
+  class Repository
+    def initialize(regions, periods:)
+      @regions = Array(regions)
+      @regions_by_id = @regions.map { |r| [r.id, r] }.to_h
+
+      @periods = if periods.is_a?(Period)
+        Range.new(periods, periods).to_a
+      else
+        periods.to_a
+      end
+    end
+
+    attr_reader :regions, :periods
+    attr_reader :regions_by_id
+    delegate :cache, :logger, to: Rails
+
+    class CacheKey
+      def initialize(key)
+        @key = key
+        @parts = key.split("/")
+      end
+
+      def region_id
+        @parts[1]
+      end
+
+      def slug
+        @parts[2]
+      end
+
+      def period
+        Period.new(type: @parts[3], value: @parts[4])
+      end
+    end
+
+    def lookup_region(id)
+      @regions_by_id.fetch(id)
+    end
+
+    def controlled_patients_info
+      keys = cache_keys(:controlled_patients_info)
+      cached_results = cache.fetch_multi(*keys) { |key|
+        cache_key = CacheKey.new(key)
+        period = cache_key.period
+        region = lookup_region(cache_key.region_id)
+        controlled_patients_for(region, period).count
+      }
+      cached_results.each_with_object({}) do |(key, count), results|
+        cache_key = CacheKey.new(key)
+        results[cache_key.slug] ||= Hash.new(0)
+        results[cache_key.slug][cache_key.period] = count
+      end
+    end
+
+    def uncontrolled_patients_info
+      keys = cache_keys(:uncontrolled_patients_info)
+      cached_results = cache.fetch_multi(*keys) { |key|
+        cache_key = CacheKey.new(key)
+        period = cache_key.period
+        region = lookup_region(cache_key.region_id)
+        uncontrolled_patients_for(region, period).count
+      }
+      cached_results.each_with_object({}) do |(key, count), results|
+        cache_key = CacheKey.new(key)
+        results[cache_key.slug] ||= Hash.new(0)
+        results[cache_key.slug][cache_key.period] = count
+      end
+    end
+
+    def cache_keys(operation)
+      combinations = regions.to_a.map(&:cache_key_v2).product(periods.map(&:cache_key))
+      combinations.map { |region_key, period_key| [region_key, period_key, operation].join("/") }
+    end
+
+    def controlled_patients_for(region, period)
```

Comment:
All the below should probably be extracted to a new Query object, so that we can consolidate the logic from here and from the ControlRateService.

Reply:
Been thinking about this for a while: We should port the query objects to support the new `Period` interface.

Reply:
yeah, the old query has a few problems that prevented me from reusing it:

* it doesn't support regions
* it doesn't support the Period model
* it assumes you want to start from current period and always retrieve the last n periods
* it applies a `recorded_at` where clause to monthly_patients that we don't want 

I think we should extract the new queries for reuse and move away from the old ones.